### PR TITLE
avoid suppressing errors in deferred funcs

### PR DESCRIPTION
**Pull request by mdb**
This seeks to address issue #129 by exiting nonzero if ever the `defer`'d `switchBackToRemoteFunc` funcs return an error.

By using a closure in concert with `errors.Join`, we can join an error returned by the `defer`'d func with any other error(s), surface visibility on all errors, and avoid exiting successfully if ever an error occurs within the `defer`'d func.

This is especially helpful in CI/CD pipelines where non-successful `tfmigrate plan` execution likely indicates a problem that should gate the execution of the pipeline.

@minamijoyo Is this a technique you'd be receptive to? If so, I can give some more thoughts to acceptance testing it.

### CODE REVIEW

**Comment by mdb** [comment]
In the absence of more sophisticated tests, I've been using the following shell script and observing that it exits non-zero:

```bash
#!/bin/bash

# compile a fresh bin/tfmigrate
make build

# stop and remove containers from previous runs, just to make sure we're testing
# in a fresh "clean room" environment
docker-compose down --remove-orphans

# start localstack
docker-compose up -d localstack
docker-compose run --rm dockerize -wait tcp://localstack:4566 -timeout 60s

# make sure the tfstate-test s3 bucket exists
docker-compose exec -T localstack /etc/localstack/init/wait_s3_bucket_exists.sh

# enter the test-fixtures/legacy-tfstate directory
cd test-fixtures/legacy-tfstate

# remove files from previous runs that may impact the test
rm terraform.tfstate || true
rm -rf .terraform || true
rm -rf terraform.tf || true
rm -rf .terraform.lock.hcl || true

# add a terraform.tf file with a backend configuration that points to localstack
cat <<EOT >> terraform.tf
terraform {
  # https://www.terraform.io/docs/backends/types/s3.html
  backend "s3" {
    region = "ap-northeast-1"
    # NOTE that tfstate-test is commented out.
    #bucket = "tfstate-test"
    key    = "test/terraform.tfstate"

    # mock s3 endpoint with localstack
    endpoint                    = "http://localhost:4566"
    access_key                  = "dummy"
    secret_key                  = "dummy"
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    force_path_style            = true
  }
}
EOT

# initialize and apply the configuration using TF 0.12.31
# This seeds the tfstate-test bucket with a terraform.tfstate file
tfenv use 1.4.4
terraform init
terraform plan
terraform apply -auto-approve

# Remove the .terraform directory so that we can test the migration from
# scratch, just as we would in a real-world CI/CD-orchestrated scenario.
# rm -rf .terraform

# Make sure there are no migration.hcl files from previous runs of this script.
rm migration.hcl || true

# Create a fresh migration.hcl file.
cat <<EOT >> migration.hcl
migration "state" "test" {
  dir = "."
  actions = [
    "replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null",
  ]
}
EOT

# Attempt to plan the migration.
../../bin/tfmigrate plan migration.hcl
```

**Comment by minamijoyo** [comment]
Thank you for your suggestion! Using named return values is a good idea.

For testing, I don't think it's necessary to check from outside the CLI. How about testing Migrator.Plan() as follows?
- If rollback fails inside the deferred function with an invalid backend configuration, such as #129, then Plan() returns an error.
- If state commands fail in actions with an invalid migration file, we can use more simple cases without a legacy state, such as invalid addresses for state mv, then Plan() returns an error.
- If both occur at the same time, then Plan() returns an error, and both errors are joined.

What do you think?

**Comment by mdb** [comment]
@minamijoyo IIUC, I think that sounds reasonable - I've pushed a commit doing so. Did I interpret your suggestion correctly?

File: tfmigrate/multi_state_migrator.go [hunk 1]
```diff
@@ -182,14 +188,17 @@ func (m *MultiStateMigrator) plan(ctx context.Context) (*tfexec.State, *tfexec.S
 					log.Printf("[ERROR] [migrator@%s] unexpected diffs\n", m.toTf.Dir())
 					return nil, nil, fmt.Errorf("terraform plan command returns unexpected diffs in %s to_dir: %s", m.toTf.Dir(), err)
 				}
+
 				log.Printf("[INFO] [migrator@%s] unexpected diffs, ignoring as force option is true: %s", m.toTf.Dir(), err)
+				// reset err to nil to intentionally ignore unexpected diffs.
+				err = nil
```
Comment (minamijoyo) [inline tfmigrate/multi_state_migrator.go:194]: We also need to reset the err to nil in the from dir at line 172.

**Comment by minamijoyo** [comment]
@mdb Thank you for working on this! It almost looks good. I left one comment about resetting the error; please check it out.

File: tfmigrate/multi_state_migrator.go [hunk 2]
```diff
@@ -169,6 +169,8 @@ func (m *MultiStateMigrator) plan(ctx context.Context) (fromCurrentState *tfexec
 					log.Printf("[ERROR] [migrator@%s] unexpected diffs\n", m.fromTf.Dir())
 					return nil, nil, fmt.Errorf("terraform plan command returns unexpected diffs in %s from_dir: %s", m.fromTf.Dir(), err)
 				}
+				// reset err to nil to intentionally ignore unexpected diffs.
+				err = nil
```
Comment (minamijoyo) [inline tfmigrate/multi_state_migrator.go:173]: @mdb My bad. I said line 172, but we should reset the error after logging it.

**Review by minamijoyo — APPROVED** [review]
LGTM 👍

**Comment by minamijoyo** [comment]
@mdb Thank you for the fix! I've cut a new release v0.3.16 🚢
